### PR TITLE
[RPA-1479] Tenants backend implementation

### PR DESCRIPTION
**Pull request by Aeri4a**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

File: runbotics-orchestrator/src/main/java/com/runbotics/service/UserService.java [hunk 1]
```diff
@@ -200,6 +206,11 @@ public User createUser(AdminUserDTO userDTO) {
                 .collect(Collectors.toSet());
             user.setAuthorities(authorities);
         }
+
+        Tenant defaultTenant = new Tenant();
+        defaultTenant.setId(UUID.fromString("b7f9092f-5973-c781-08db-4d6e48f78e98"));
```
Comment (star-in-the-jar) [inline runbotics-orchestrator/src/main/java/com/runbotics/service/UserService.java]: in lines 163-164 there's the same. Create function getDefaultTenant

File: runbotics-orchestrator/src/main/java/com/runbotics/service/dto/TenantDTO.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        if (!(o instanceof TenantDTO)) {
+            return false;
+        }
+        return id != null && id.equals(((TenantDTO) o).id);
+    }
+
+    @Override
+    public int hashCode() {
+        return getClass().hashCode();
+    }
```
Comment (star-in-the-jar) [inline runbotics-orchestrator/src/main/java/com/runbotics/service/dto/TenantDTO.java]: what is it for?

File: runbotics-orchestrator/src/main/java/com/runbotics/service/impl/GuestServiceImpl.java [hunk 3]
```diff
@@ -97,6 +98,10 @@ private User createGuestUser(String langKey) {
         guestUser.setCreatedBy("system");
         guestUser.setActivationKey(RandomUtil.generateActivationKey());
 
+        Tenant defaultTenant = new Tenant();
+        defaultTenant.setId(UUID.fromString("b7f9092f-5973-c781-08db-4d6e48f78e98"));
```
Comment (star-in-the-jar) [inline runbotics-orchestrator/src/main/java/com/runbotics/service/impl/GuestServiceImpl.java]: again -> use getDefaultTenant here

or even in class tenant assign default id and create a constructor that takes id as parameter - your choice

File: runbotics-orchestrator/src/main/java/com/runbotics/web/rest/admin/TenantAdminResource.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    @DeleteMapping("/{id}")
+    public ResponseEntity<Void> deleteTenant(@PathVariable(value = "id") final UUID id) {
+        log.debug("REST request to delete tenant by id: {}", id);
+        tenantService.delete(id);
+
+        return ResponseEntity
+            .noContent()
+            .headers(HeaderUtil.createEntityDeletionAlert(applicationName, true, ENTITY_NAME, id.toString()))
+            .build();
+    }
```
Comment (star-in-the-jar) [inline runbotics-orchestrator/src/main/java/com/runbotics/web/rest/admin/TenantAdminResource.java:128]: what about checking the permissions? FKs

File: runbotics-orchestrator/src/main/java/com/runbotics/web/rest/tenant/TenantResource.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+//    public ResponseEntity<TenantDTO> updateUserTenant(@Valid @RequestBody TenantDTO tenantDTO) {
+//        log.debug("REST request to partial update tenant by requester");
+//        Optional<TenantDTO> updatedTenant = tenantService.partialUpdate(tenantDTO);
+//
+//        return ResponseUtil.wrapOrNotFound(
+//            updatedTenant,
+//            HeaderUtil.createEntityUpdateAlert(applicationName, true, ENTITY_NAME, tenantDTO.getId().toString())
+//        );
+//    }
+}
```
Comment (star-in-the-jar) [inline runbotics-orchestrator/src/main/java/com/runbotics/web/rest/tenant/TenantResource.java:54]: add short comment above the commented code why is it left

File: runbotics/runbotics-common/src/model/api/authority.model.ts [hunk 6]
```diff
@@ -2,6 +2,7 @@ import { IFeatureKey } from "./feature-key.model";
 
 export enum Role {
     ROLE_ADMIN = 'ROLE_ADMIN',
+    ROLE_TENANT_ADMIN = 'ROLE_TENANT_ADMIN',
```
Comment (star-in-the-jar) [inline runbotics/runbotics-common/src/model/api/authority.model.ts]: can be left commented for now

File: runbotics/runbotics-common/src/model/api/feature-key.model.ts [hunk 7]
```diff
@@ -183,6 +183,15 @@ export enum FeatureKey {
...
     PROCESS_COLLECTION_DELETE = 'PROCESS_COLLECTION_DELETE',
+
+    // Read/write access to every tenant resource
+    TENANT_ALL_ACCESS = 'TENANT_ALL_ACCESS',
+
+    // Allows to read the resources assigned to the tenant
+    TENANT_READ = 'TENANT_READ',
+
+    // Allows to edit the resources assigned to the tenant
+    TENANT_EDIT = 'TENANT_EDIT',
```
Comment (star-in-the-jar) [inline runbotics/runbotics-common/src/model/api/feature-key.model.ts:197]: what about TENANT_DELTE

**Review by star-in-the-jar — CHANGES_REQUESTED** [review]
+ version bump

File: runbotics-orchestrator/src/main/java/com/runbotics/service/dto/TenantDTO.java (hunk 2 above)
Comment (Aeri4a) [reply runbotics-orchestrator/src/main/java/com/runbotics/service/dto/TenantDTO.java]: I think it isn't. My bad.

File: runbotics-orchestrator/src/main/java/com/runbotics/service/UserService.java [hunk 8]
```diff
@@ -157,6 +159,9 @@ public User registerUser(AdminUserDTO userDTO, String password) {
         Set<Authority> authorities = new HashSet<>();
         authorityRepository.findById(AuthoritiesConstants.USER).ifPresent(authorities::add);
         newUser.setAuthorities(authorities);
+
+        newUser.setTenant(Utils.getDefaultTenant());
```
Comment (star-in-the-jar) [inline runbotics-orchestrator/src/main/java/com/runbotics/service/UserService.java:164]: 👍

**Review by star-in-the-jar — APPROVED** [review]
+ version bump if it's not done yet

**Pull request merged**